src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,123 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int ARRAY_START_SIZE = 10;
+    private Object[] list;
+    private int size;
+
+    public ArrayList() {
+        list = new Object[ARRAY_START_SIZE];
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        growIfArrayFull();
+        list[size()] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexIsValid(index, true);
+        growIfArrayFull();
+        Object[] newList = new Object[list.length];
+        if (index == size()) {
+            add(value);
+        } else {
+            if (index >= 0) {
+                System.arraycopy(list, 0, newList, 0, index);
+            }
+            newList[index] = value;
+            if (size() - index >= 0) {
+                System.arraycopy(list, index, newList, index + 1, size() - index);
+            }
+            list = newList;
+            size++;
+        }
     }
 
     @Override
-    public void addAll(List<T> list) {
+    public void addAll(List<T> otherList) {
+        for (int i = 0; i < otherList.size(); i++) {
+            add(otherList.get(i));
+        }
 
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexIsValid(index, false);
+        return (T) list[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndexIsValid(index, false);
+        list[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexIsValid(index, false);
+        final Object oldElement = list[index];
+        for (int i = index; i < size() - 1; i++) {
+            list[i] = list[i + 1];
+        }
+        list[size() - 1] = null;
+        size--;
+        return (T) oldElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        boolean removed = false;
+        Object[] newList = new Object[list.length];
+        int newIndex = 0;
+
+        for (int i = 0; i < size(); i++) {
+            if (!removed && (get(i) != null ? get(i).equals(element) : get(i) == element)) {
+                removed = true;
+                size--;
+                continue;
+            }
+            newList[newIndex++] = list[i];
+        }
+        list = newList;
+
+        if (!removed) {
+            throw new NoSuchElementException("No such element " + element);
+        }
```

Comment:
you can simplify it. Just find the index of the element and call the method remove by index.